src/test/resources/refaster/PreConditionsVerifierRecipes.java

```diff
@@ -0,0 +1,403 @@
+
+        @Override
+        public String getDisplayName() {
+            return "Refaster template `PreConditionsVerifier.UsesTypeMapOrListWhenBeforeTemplateContainsMapAndList`";
+        }
+
+        @Override
+        public String getDescription() {
+            return "Recipe created for the following Refaster template:\n```java\npublic static class UsesTypeMapOrListWhenBeforeTemplateContainsMapAndList {\n    \n    @BeforeTemplate()\n    void before(List<?> actual) {\n        System.out.println(actual);\n    }\n    \n    @BeforeTemplate()\n    void beforeTwo(Map<?, ?> actual) {\n        System.out.println(actual);\n    }\n    \n    @AfterTemplate()\n    void after(Object actual) {\n        System.out.println(\"Changed: \" + actual);\n    }\n}\n```\n.";
+        }
+
+        @Override
+        public TreeVisitor<?, ExecutionContext> getVisitor() {
+            JavaVisitor<ExecutionContext> javaVisitor = new AbstractRefasterJavaVisitor() {
+                final JavaTemplate before = JavaTemplate
+                        .builder("System.out.println(#{actual:any(java.util.List<?>)});")
+                        .build();
+                final JavaTemplate beforeTwo = JavaTemplate
+                        .builder("System.out.println(#{actual:any(java.util.Map<?,?>)});")
+                        .build();
+                final JavaTemplate after = JavaTemplate
+                        .builder("System.out.println(\"Changed: \" + #{actual:any(java.lang.Object)});")
+                        .build();
+
+                @Override
+                public J visitMethodInvocation(J.MethodInvocation elem, ExecutionContext ctx) {
+                    JavaTemplate.Matcher matcher;
+                    if ((matcher = before.matcher(getCursor())).find()) {
+                        maybeRemoveImport("java.util.List");
+                        return embed(
+                                after.apply(getCursor(), elem.getCoordinates().replace(), matcher.parameter(0)),
+                                getCursor(),
+                                ctx,
+                                SHORTEN_NAMES
+                        );
+                    }
+                    if ((matcher = beforeTwo.matcher(getCursor())).find()) {
+                        maybeRemoveImport("java.util.Map");
+                        return embed(
+                                after.apply(getCursor(), elem.getCoordinates().replace(), matcher.parameter(0)),
+                                getCursor(),
+                                ctx,
+                                SHORTEN_NAMES
+                        );
+                    }
+                    return super.visitMethodInvocation(elem, ctx);
+                }
+
+            };
+            return Preconditions.check(
+                    Preconditions.and(
+                            new UsesMethod<>("java.io.PrintStream println(..)", true),
+                            Preconditions.or(
```

Comment:
Name: **UsesTypeMapOrListWhenBeforeTemplateContainsMapAndListRecipe**, thus:
- Use `new UsesType<>("java.util.List", true)` OR `new UsesType<>("java.util.Map", true)`

Reply:
In this case I'd expect no change to the preconditions either, as we already have that `or` between the method and types in use.